bigtable/client/row_range.cc

```diff
@@ -0,0 +1,92 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "bigtable/client/row_range.h"
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+namespace btproto = ::google::bigtable::v2;
+
+bool RowRange::IsEmpty() const {
+  std::string unused;
+  // We do not want to copy the strings unnecessarily, so initialize a reference
+  // pointing to *_key_closed() or *_key_open(), as needed.
+  std::reference_wrapper<const std::string> start(unused);
+  bool start_open = false;
+  switch (row_range_.start_key_case()) {
+    case btproto::RowRange::kStartKeyClosed:
+      start = std::cref(row_range_.start_key_closed());
+      break;
+    case btproto::RowRange::kStartKeyOpen:
+      start = std::cref(row_range_.start_key_open());
+      start_open = true;
+      break;
+    case btproto::RowRange::START_KEY_NOT_SET:
+      // A range starting at -infinity is never empty.
+      return false;
+  }
+  std::reference_wrapper<const std::string> end(unused);
+  bool end_open = false;
+  switch (row_range_.end_key_case()) {
```

Comment:
Does this automatically handle missing enum values from the `switch`? Do we need a `default` clause here and elsewhere (e.g., if a new enum is added in the future)?

Reply:
"Automatically" as it will not compile and force us to fix it: `end_key_case()` returns an enum with only the three values in this `switch()` statement.  Some of the compilers we use issue warnings when you miss an `enum` value, and because we compile with `-Werror` we will break the build.  Which is a good thing I believe.